File: python/phone-number/phone_number.py
"""
Validate if a given string is a proper phone number and if so return it in the desired format
"""

import re

class PhoneNumber:
    """
    Phone number class
    """
    def __init__(self, number):
        """
        Check if a number is a proper phone number.
        If valid, parse out the sections and and save it
        to the property .number (just the digits).
        """
        phone_number_regex = r'(\+?1[ -\.]*)?\(?(?P<area_code>[0-9]{3})\)?[ -\.]*(?P<exchange>[0-9]{3})[ -\.]*(?P<subscriber>[0-9]{4})'
        symbols = [symbol for symbol in number if symbol in r'!@#$%^&*;:<>,?/\|']
        if len(symbols) > 0:
            raise ValueError('punctuations not permitted')
        letters = [letter for letter in number if ('a' <= letter  <= 'z') or ('A' <= letter <= 'Z')]
        if len(letters) > 0:
            raise ValueError('letters not permitted')
        digits = [digit for digit in number if '0' <= digit <= '9']
        if len(digits) < 10:
            raise ValueError('must not be fewer than 10 digits')
        if len(digits) > 11:
            raise ValueError('must not be greater than 11 digits')
        if len(digits) == 11 and digits[0] != '1':
            raise ValueError('11 digits must start with 1')

        result = re.search(phone_number_regex, number)
        self.number = ''
        if result is not None:
            self.area_code = result.group('area_code') 
            if self.area_code[0] == '0':
                raise ValueError('area code cannot start with zero')
            if self.area_code[0] == '1':
                raise ValueError('area code cannot start with one')
                
            self.exchange = result.group('exchange') 
            if self.exchange[0] == '0':
                raise ValueError('exchange code cannot start with zero')            
            if self.exchange[0] == '1':
                raise ValueError('exchange code cannot start with one')
            
            self.subscriber = result.group('subscriber')
            self.number = self.area_code + self.exchange + self.subscriber

    def pretty(self):
        """
        Pretty print the class' phone number
        :returns: Pretty version of the class' phone number
        """
        return f'({self.area_code})-{self.exchange}-{self.subscriber}'

File: go/phone-number/phone_number.go
package phonenumber

import (
	"errors"
	"fmt"
	"regexp"
)

// Parse the give phone number using a regular expression. Return the find matches result.
// @param phoneNumber: The phone number to parse
// @returns: error if regex did not find a number else an array of strings with the first entry
// being the entire group matched followed by named matches in order.
func ParsePhoneNumber(phoneNumber string) ([]string, error) {
	pattern := `^\+?1?[ \.-]*\(?(?<area_code>[2-9]\d{2})\)?[ \.-]*(?<exchange>[2-9]\d{2})[ \.-]*(?<subscriber>\d{4})[ \.-]*$`
	re := regexp.MustCompile(pattern)
	matches := re.FindStringSubmatch(phoneNumber)
	if matches == nil {
		return make([]string, 0), errors.New("Phone number not found.")
	}
	return matches, nil
}

// Return the phone number (area code, exchange, and subscriber as a string of all digits)
// from the given phoneNumber parameter
// @param phoneNumber: The phone number to parse
// @returns error if phone number not parsed otherwise area code, exchange and subscriber as one long string of digits
func Number(phoneNumber string) (string, error) {
	parts, err := ParsePhoneNumber(phoneNumber)
	if err != nil {
		return "", err
	}
	return parts[1] + parts[2] + parts[3], nil
}

// Return the area code of a phone number (just the digits) from the given phoneNumber parameter
// @param phoneNumber: The phone number to parse
// @returns error if phone number not parsed otherwise the digits of the area code as a string
func AreaCode(phoneNumber string) (string, error) {
	parts, err := ParsePhoneNumber(phoneNumber)
	if err != nil {
		return "", err
	}
	return parts[1], nil
}

// Parse and return a phone number in the desired format (NNN) NNN-NNNN (N is a digit)
// @param phoneNumber: The phone number to parse
// @returns error is phone number not parsed otherwise a formatted phone number string.
func Format(phoneNumber string) (string, error) {
	number, err := Number(phoneNumber)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("(%s) %s-%s", number[:3], number[3:6], number[6:]), nil
}
